pipelineRunner.go

```diff
@@ -50,712 +49,527 @@ func NewPipelineRunner(envvarHelper EnvvarHelper, whenEvaluator WhenEvaluator, d
 
 	span, ctx := opentracing.StartSpanFromContext(ctx, "RunStage")
 	defer span.Finish()
-	span.SetTag("stage", p.Name)
+	span.SetTag("stage", stage.Name)
 
 	parentStageName := ""
+	stagePlaceholder := fmt.Sprintf("[%v]", stage.Name)
 	if parentStage != nil {
 		parentStageName = parentStage.Name
+		stagePlaceholder = fmt.Sprintf("[%v] [%v]", parentStageName, stage.Name)
 	}
 
-	p.ContainerImage = os.Expand(p.ContainerImage, pr.envvarHelper.getEstafetteEnv)
+	log.Info().Msgf("%v Starting stage '%v'", stagePlaceholder, stage.Name)
+	var isPulledImage bool
+	var isTrustedImage bool
+	var imagePullDuration time.Duration
+	var imageSize int64
+	var buildLogStepDockerImage *contracts.BuildLogStepDockerImage
 
-	result.Stage = p
-	result.Depth = depth
-	result.RunIndex = runIndex
-	result.LogLines = make([]contracts.BuildLogLine, 0)
+	if stage.ContainerImage != "" {
+		stage.ContainerImage = os.Expand(stage.ContainerImage, pr.envvarHelper.getEstafetteEnv)
+		isPulledImage = pr.dockerRunner.IsImagePulled(stage.Name, stage.ContainerImage)
+		isTrustedImage = pr.dockerRunner.IsTrustedImage(stage.Name, stage.ContainerImage)
 
-	log.Info().Msgf("[%v] Starting stage '%v'", p.Name, p.Name)
+		if !isPulledImage || runtime.GOOS == "windows" {
 
-	if p.ContainerImage != "" {
-		result.IsDockerImagePulled = pr.dockerRunner.isDockerImagePulled(p.Name, p.ContainerImage)
-		result.IsTrustedImage = pr.dockerRunner.isTrustedImage(p.Name, p.ContainerImage)
-
-		if !result.IsDockerImagePulled || runtime.GOOS == "windows" {
-
-			// log tailing - start stage as pending
-			status := "PENDING"
-			pr.tailLogsChannel <- contracts.TailLogLine{
-				Step:         p.Name,
-				ParentStage:  parentStageName,
-				Type:         "stage",
-				Depth:        depth,
-				RunIndex:     runIndex,
-				Image:        getBuildLogStepDockerImage(result),
-				AutoInjected: &result.Stage.AutoInjected,
-				Status:       &status,
+			buildLogStepDockerImage = &contracts.BuildLogStepDockerImage{
+				Name:      getContainerImageName(stage.ContainerImage),
+				Tag:       getContainerImageTag(stage.ContainerImage),
+				IsTrusted: isTrustedImage,
+				IsPulled:  isPulledImage,
 			}
 
+			// start pulling stage
+			pr.sendStatusMessage(stage, parentStageName, depth, runIndex, buildLogStepDockerImage, nil, contracts.StatusPending)
+
 			// pull docker image
 			dockerPullStart := time.Now()
-			result.DockerPullError = pr.dockerRunner.runDockerPull(ctx, p.Name, p.ContainerImage)
-			result.DockerPullDuration = time.Since(dockerPullStart)
-			if result.DockerPullError != nil {
-				return result, result.DockerPullError
-			}
+			err = pr.dockerRunner.PullImage(ctx, stage.Name, stage.ContainerImage)
+			imagePullDuration = time.Since(dockerPullStart)
 		}
 
 		// set docker image size
-		size, err := pr.dockerRunner.getDockerImageSize(p.ContainerImage)
-		if err != nil {
-			return result, err
+		if err == nil {
```

Comment:
err can only contain something inside the if block above, right?
so maybe it's better to return if there is an error on the pull image phase

Reply:
Ah, need to make sure there's a size variable ahead, so i can set with `=` instead of `:=`. I deliberately not return errors directly, because the status needs to be set via the taillog; it would litter the code with statements for that in that case.  Looking forward to golangs option to `catch` and handle all errors in a func so it can be returned immediately.

Reply:
Oh, I actually already use `=` everywhere that `err` is assigned a value; This particular check happens because `PullImage` could have returned an error in which case we still want the part with `	finalStatus` below to run.

Reply:
ok, got it now

Reply:
I changed it quite a lot now so we can return immediately, by having a defer statement at the top of the function.